### PR TITLE
18752 base-address upgrade

### PR DESCRIPTION
**Pull request by PaulGarewal**
*Issue #:* /bcgov/entity#18752

*Description of changes:*

- Added new vue3 base address from Assets Team. 
- Testing still needed to be completed.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

**Comment by PaulGarewal** [comment]
Not sure if this is what we are looking for:

image

File: src/components/base-address/package.json
Comment (severinbeauvais) [inline src/components/base-address/package.json:3]: Don't update this -- the publish step (done after your PR is merged) handles this.

Also, we'll want to change the major version number!

File: src/components/base-address/interfaces/address-interface.ts
Comment (severinbeauvais) [inline src/components/base-address/interfaces/address-interface.ts:9]: Is this an internal-only interface? If not then it should be in the interfaces folder so that UIs can import it. (There may be an address interface there already. Can it be reused?)

**Comment by severinbeauvais** [comment]
You will know you have succeeded when this component can be used externally and works properly there. Hopefully it can be used in the same way as the old component. You may need to create a very simple Vue3 UI to test this.

**Review by JazzarKarim — APPROVED** [review]
LGTM as long as you've tested it Paul 👍

**Comment by severinbeauvais** [comment]
> Not sure if this is what we are looking for (but it is a start):

So there's no labels on the fields? Ask UI if we can use it like this (which is presumably the same as in the Assets/PPR UI).

**Comment by PaulGarewal** [comment]
> > Not sure if this is what we are looking for (but it is a start):
> 
> So there's no labels on the fields? Ask UI if we can use it like this (which is presumably the same as in the Assets/PPR UI).

Ok will do. Here is what the fields look like before the validation:

image

**Comment by PaulGarewal** [comment]
> > > Not sure if this is what we are looking for (but it is a start):
> > 
> > 
> > So there's no labels on the fields? Ask UI if we can use it like this (which is presumably the same as in the Assets/PPR UI).
> 
> Ok will do. I will ask Yui.
> 
> Here is what the fields look like before entering any values (for reference):
> 
> image

Confirmed with Yui that we want the same setup as the Assets Team for the BaseAddress.

**Comment by severinbeauvais** [comment]
> Confirmed with Yui that we want the same setup as the Assets Team for the BaseAddress.

Nice! I hoped this would be the case but now it's official.

**Comment by severinbeauvais** [comment]
> You will know you have succeeded when this component can be used externally and works properly there. Hopefully it can be used in the same way as the old component. You may need to create a very simple Vue3 UI to test this.

Any change this will work in a Vue2 UI?

Can you do an address lookup in Storybook? (I know it needs an API key, and we won't want to put that key in any commit in GH as our repos are public!)

Are the props and events the same as the old address component?

**Comment by cameron-eyds** [comment]
@severinbeauvais @PaulGarewal 
If your keen on testing in a Vue3 instance before committing, feel free to pull our feature branch:
https://github.com/bcgov/ppr/tree/Vue3-Upgrade-Feature-Branch

It's pretty stable, for the most part :) 
Hit me up for an .env if you want to run it. 
OR if you want to commit and give it a go as a true import, the above holds true as well 👍

File: package.json
Comment (severinbeauvais) [inline package.json:17]: Since you're changing this file, please update the patch version on line 3.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:1]: I think you're missing "base-address/index.ts".

File: src/components/base-address/enums/validation-rule.ts
Comment (severinbeauvais) [inline src/components/base-address/enums/validation-rule.ts:1]: Is this internal only, or will an app need to import this?

If so then you have 2 options:
1. export everything needed in this "base-address/index.ts"
2. move this to shared enums (maybe with a more specific name like address-validation-rules.ts)

We haven't used option 1 yet in this repo but it feels like a good idea to keep all related things together. What do you think?

File: src/components/base-address/factories/countries-provinces-factory.ts
Comment (severinbeauvais) [inline src/components/base-address/factories/countries-provinces-factory.ts:2]: You need a "base-address/package.json" (see other components for examples; set version to 0.0.0 initially) to specify all this component's dependencies.

Check the other files in here for additional dependencies.

File: package.json
Comment (severinbeauvais) [reply package.json:17]: Wait - this file is used by Storybook. Does SB need this package, or is it needed by the component only?

File: src/components/base-address/BaseAddress.vue
Comment (PaulGarewal) [reply src/components/base-address/BaseAddress.vue:1]: It is there:

image

File: src/interfaces/addresses-interface.ts
Comment (severinbeauvais) [inline src/interfaces/addresses-interface.ts:7]: Move this last, and indent it.

(It's a cheat to allow undeclared properties in this interface/type. Undesirable but we're not going to fix this right now.)

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: Oops, missed it. Sorry and thanks!

Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: If an app wants to use this component, and it needs some enums/interfaces/etc that we decided to keep in this folder, then those should be export above as well.

File: src/components/base-address/enums/validation-rule.ts
Comment (PaulGarewal) [reply src/components/base-address/enums/validation-rule.ts:1]: Option 2 it is 👍

File: src/components/base-address/factories/countries-provinces-factory.ts
Comment (PaulGarewal) [reply src/components/base-address/factories/countries-provinces-factory.ts:2]: This is the current base-address/package.json:
image
Should I reset the version to 0.0.0 and add the relevant dependencies?

File: src/components/base-address/factories/validation-factory.ts
Comment (severinbeauvais) [inline src/components/base-address/factories/validation-factory.ts:3]: This should be `@bcrs-shared-components/enums` otherwise here (ie, when Storybook uses this component), it resolves to the local enums folder, but when used in an app, it will resolve to the app's enums folder.

Comment (severinbeauvais) [reply src/components/base-address/factories/validation-factory.ts:3]: This is a problem here: https://github.com/bcgov/bcrs-shared-components/blob/4c62c41230422c5ceef744176fdc1bb04d6b5372/src/components/date-picker/DatePicker.vue#L84

But Leo said he got it working: https://github.com/bcgov/bcrs-shared-components/pull/215#discussion_r1417854456

File: src/enums/address-validation-rules.ts
Comment (severinbeauvais) [inline src/enums/address-validation-rules.ts:1]: Ideally the enum name should match the filename.

Also, our convention is that enums are plural.

File: src/components/base-address/factories/countries-provinces-factory.ts
Comment (severinbeauvais) [reply src/components/base-address/factories/countries-provinces-factory.ts:2]: Oh right, you're updating an existing component.

Leave the version alone and I'll bump it to 3.0.0 when I publish this.

File: package.json
Comment (severinbeauvais) [reply package.json:17]: Yes, scope this import to the component itself.

File: src/components/base-address/interfaces/address-interface.ts
Comment (severinbeauvais) [reply src/components/base-address/interfaces/address-interface.ts]: Fixed, thanks!

File: src/components/base-address/factories/validation-factory.ts
Comment (severinbeauvais) [inline src/components/base-address/factories/validation-factory.ts]: If you've add address-validation-rules to enums/index.ts then you can just import the "folder". See how other enums/interfaces are exported/imported.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:1]: I think you need to update line 172 also.

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.vue:246]: Do you need to import AddressValidationRules ?

File: src/components/base-address/factories/validation-factory.ts
Comment (severinbeauvais) [reply src/components/base-address/factories/validation-factory.ts]: Fixed, thanks!

File: src/components/base-address/factories/validation-factory.ts
Comment (severinbeauvais) [reply src/components/base-address/factories/validation-factory.ts]: Beautiful, thanks!

What about the 2 interfaces?

File: src/components/base-address/BaseAddress.vue
Comment (PaulGarewal) [reply src/components/base-address/BaseAddress.vue:246]: It's not valid anymore, attempting to fix

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: Again, `@/` resolves to the local interfaces folder, which will work for Storybook (the local UI, sort of) but won't work when this component is imported into an app (unless the app's local interfaces folder also has these interfaces but that's undesirable).

I think these 2 interfaces should be in the shared interfaces folder and exported along with the other (unless they're absolutely only ever used internally).

File: package.json
Comment (severinbeauvais) [inline package.json:3]: Since you're no longer changing this file, please revert the version change 😛

File: package-lock.json
Comment (severinbeauvais) [inline package-lock.json:4]: Merge issue. Delete this duplicate and run `npm i` again to fix it.

File: src/components/base-address/factories/validation-factory.ts
Comment (severinbeauvais) [reply src/components/base-address/factories/validation-factory.ts]: @PaulGarewal Should the 2 interfaces be in the shared folder?

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:246]: Is this fixed?

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: I was thinking that the imports on lines 172 (and 173) would be from the shared versions.

File: src/components/base-address/BaseAddress.stories.ts
Comment (severinbeauvais) [inline src/components/base-address/BaseAddress.stories.ts:3]: Line 3 should be from shared component (I think).

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: Also elsewhere in the vue and ts files in here.

File: src/components/base-address/resources/default-schema.ts
Comment (severinbeauvais) [inline src/components/base-address/resources/default-schema.ts:3]: I'm thinking this import should be local, as in `from "./factories/validation-factory"`. This is because the factories are internal and local to this component.

The reason I'm going on about this is that [`@/` maps to the source folder in THIS repo](https://github.com/bcgov/bcrs-shared-components/blob/4c62c41230422c5ceef744176fdc1bb04d6b5372/vite.config.ts#L15), but when these components are imported (as source files) in a UI, the UI's source folder won't have the same files. Therefore, these components should import from the "shared version" of enums or interfaces or whatever they need. Then these components should work correctly in this repo (ie, in Storybook and unit tests) and also in whatever UI they're used in.

cc: @leodube-aot

File: src/components/base-address/BaseAddress.vue
Comment (severinbeauvais) [reply src/components/base-address/BaseAddress.vue:1]: What error?
Your component should work in Storybook, both locally and after merge.

Comment (PaulGarewal) [reply src/components/base-address/BaseAddress.vue:1]: image

File: src/components/base-address/resources/default-schema.ts
Comment (leodube-aot) [reply src/components/base-address/resources/default-schema.ts]: This import should probably be `from '../factories/validation-factory'`. That resolved the error you are referencing Paul when I tested out your PR locally.

File: src/components/base-address/BaseAddress.vue
Comment (leodube-aot) [reply src/components/base-address/BaseAddress.vue:1]: See my comment below about fixing the`validation-factory` import. Once that is fixed you should be okay to ignore these import errors, as the component still runs in storybook fine for me. It looks to me like the IDE can't resolve the import, but is able to at compile/run.

File: src/components/base-address/resources/default-schema.ts
Comment (PaulGarewal) [reply src/components/base-address/resources/default-schema.ts]: Resolved the error and I ignored the other import errors. Storybook still runs now without any problems 👍 Thank you Leo.

**Review by severinbeauvais — APPROVED** [review]
LGTM!

If this component works correctly in Storybook locally (except for lookup, as previously discussed) then go ahead and merge this, and then I'll publish it.

**Review by leodube-aot — APPROVED** [review]
Good work on this Paul!

**Pull request merged**
